File: src/internal/compiler/lexer/lexer.go
package lexer

type TokenType string

const (
	TOKEN_RENDER    TokenType = "RENDER"
	TOKEN_FUNCTION  TokenType = "FUNCTION"
	TOKEN_RETURN    TokenType = "RETURN"
	TOKEN_PRINT     TokenType = "PRINT"
	TOKEN_LET       TokenType = "LET"
	TOKEN_IMPORT    TokenType = "IMPORT"
	TOKEN_BEFORE    TokenType = "BEFORE"
	TOKEN_STRING    TokenType = "STRING_TYPE"
	TOKEN_IF        TokenType = "IF"
	TOKEN_ELSE      TokenType = "ELSE"
	TOKEN_WHILE     TokenType = "WHILE"
	TOKEN_IDENT     TokenType = "IDENT"
	TOKEN_VAR       TokenType = "VAR"
	TOKEN_LITERAL   TokenType = "LITERAL"
	TOKEN_INT       TokenType = "INT"
	TOKEN_LPAREN    TokenType = "LPAREN"
	TOKEN_RPAREN    TokenType = "RPAREN"
	TOKEN_LBRACE    TokenType = "LBRACE"
	TOKEN_RBRACE    TokenType = "RBRACE"
	TOKEN_COMMA     TokenType = "COMMA"
	TOKEN_SEMICOLON TokenType = "SEMICOLON"
	TOKEN_DOT       TokenType = "DOT"
	TOKEN_ASTERISK  TokenType = "ASTERISK"
	TOKEN_LBRACKET  TokenType = "LBRACKET"
	TOKEN_RBRACKET  TokenType = "RBRACKET"
	TOKEN_PLUS      TokenType = "PLUS"
	TOKEN_MINUS     TokenType = "MINUS"
	TOKEN_LT        TokenType = "LT"
	TOKEN_GT        TokenType = "GT"
	TOKEN_EQ        TokenType = "EQ"
	TOKEN_NOT_EQ    TokenType = "NOT_EQ"
	TOKEN_QUESTION  TokenType = "QUESTION"
	TOKEN_COALESCE  TokenType = "COALESCE"
	TOKEN_BANG      TokenType = "BANG"
	TOKEN_ASSIGN    TokenType = "ASSIGN"
	TOKEN_COLON     TokenType = "COLON"
	TOKEN_PIPE      TokenType = "PIPE"
	TOKEN_NULLSAFE  TokenType = "NULLSAFE"
	TOKEN_MATCH     TokenType = "MATCH"
	TOKEN_ARROW     TokenType = "ARROW"
	TOKEN_FN        TokenType = "FN"
	TOKEN_SPAWN     TokenType = "SPAWN"
	TOKEN_YIELD     TokenType = "YIELD"
	TOKEN_ENUM      TokenType = "ENUM"
	TOKEN_READONLY  TokenType = "READONLY"
	TOKEN_AND       TokenType = "AND"
	TOKEN_OR        TokenType = "OR"
	TOKEN_EOF       TokenType = "EOF"
	TOKEN_ILLEGAL   TokenType = "ILLEGAL"
)

type Token struct {
	Type      TokenType
	Literal   string
	Delimiter byte
}

type Lexer struct {
	input        string
	position     int
	readPosition int
	ch           byte
}

func New(input string) *Lexer {
	l := &Lexer{input: input}
	l.readChar()
	return l
}

func (l *Lexer) readChar() {
	if l.readPosition >= len(l.input) {
		l.ch = 0
	} else {
		l.ch = l.input[l.readPosition]
	}
	l.position = l.readPosition
	l.readPosition++
}

func (l *Lexer) NextToken() Token {
	var tok Token
	l.skipWhitespace()
	switch l.ch {
	case '(':
		tok = Token{Type: TOKEN_LPAREN, Literal: string(l.ch)}
	case ')':
		tok = Token{Type: TOKEN_RPAREN, Literal: string(l.ch)}
	case '{':
		tok = Token{Type: TOKEN_LBRACE, Literal: string(l.ch)}
	case '}':
		tok = Token{Type: TOKEN_RBRACE, Literal: string(l.ch)}
	case ':':
		tok = Token{Type: TOKEN_COLON, Literal: string(l.ch)}
	case ',':
		tok = Token{Type: TOKEN_COMMA, Literal: string(l.ch)}
	case '[':
		tok = Token{Type: TOKEN_LBRACKET, Literal: string(l.ch)}
	case ']':
		tok = Token{Type: TOKEN_RBRACKET, Literal: string(l.ch)}
	case '+':
		tok = Token{Type: TOKEN_PLUS, Literal: string(l.ch)}
	case '*':
		tok = Token{Type: TOKEN_ASTERISK, Literal: string(l.ch)}
	case ';':
		tok = Token{Type: TOKEN_SEMICOLON, Literal: string(l.ch)}
	case '/':
		if l.peekChar() == '/' {
			l.skipComment()
			return l.NextToken()
		}
		tok = Token{Type: TOKEN_IDENT, Literal: string(l.ch)}
	case '.':
		tok = Token{Type: TOKEN_DOT, Literal: string(l.ch)}
	case '?':
		if l.peekChar() == '?' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: TOKEN_COALESCE, Literal: string(ch) + string(l.ch)}
		} else if l.peekChar() == '-' {
			if l.peekPeekChar() == '>' {
				ch := l.ch
				l.readChar()
				l.readChar()
				tok = Token{Type: TOKEN_NULLSAFE, Literal: string(ch) + "->"}
			} else {
				tok = Token{Type: TOKEN_QUESTION, Literal: string(l.ch)}
			}
		} else {
			tok = Token{Type: TOKEN_QUESTION, Literal: string(l.ch)}
		}
	case '-':
		tok = Token{Type: TOKEN_MINUS, Literal: string(l.ch)}
	case '<':
		tok = Token{Type: TOKEN_LT, Literal: string(l.ch)}
	case '>':
		tok = Token{Type: TOKEN_GT, Literal: string(l.ch)}
	case '!':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: TOKEN_NOT_EQ, Literal: string(ch) + string(l.ch)}
		} else {
			tok = Token{Type: TOKEN_BANG, Literal: string(l.ch)}
		}
	case '|':
		if l.peekChar() == '>' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: TOKEN_PIPE, Literal: string(ch) + string(l.ch)}
		} else {
			tok = Token{Type: TOKEN_OR, Literal: string(l.ch)}
		}
	case '&':
		tok = Token{Type: TOKEN_AND, Literal: string(l.ch)}
	case '=':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: TOKEN_EQ, Literal: string(ch) + string(l.ch)}
		} else if l.peekChar() == '>' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: TOKEN_ARROW, Literal: string(ch) + string(l.ch)}
		} else {
			tok = Token{Type: TOKEN_ASSIGN, Literal: string(l.ch)}
		}
	case '$':
		tok.Type = TOKEN_VAR
		l.readChar()
		tok.Literal = l.readIdentifier()
		return tok
	case '"':
		tok.Type = TOKEN_LITERAL
		tok.Delimiter = '"'
		tok.Literal = l.readQuotedString('"')
	case '\'':
		tok.Type = TOKEN_LITERAL
		tok.Delimiter = '\''
		tok.Literal = l.readQuotedString('\'')
	case 0:
		tok.Type = TOKEN_EOF
		tok.Literal = ""
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = lookupIdent(tok.Literal)
			return tok
		} else if isDigit(l.ch) {
			tok.Type = TOKEN_INT
			tok.Literal = l.readNumber()
			return tok
		} else {
			tok = Token{Type: TOKEN_ILLEGAL, Literal: string(l.ch)}
		}
	}
	l.readChar()
	return tok
}

func (l *Lexer) readIdentifier() string {
	position := l.position
	for isLetter(l.ch) || isDigit(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func (l *Lexer) readQuotedString(delimiter byte) string {
	position := l.position + 1
	for {
		l.readChar()
		if l.ch == '\\' {
			l.readChar()
			continue
		}
		if l.ch == delimiter || l.ch == 0 {
			break
		}
	}
	return l.input[position:l.position]
}

func (l *Lexer) peekChar() byte {
	if l.readPosition >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition]
}

func (l *Lexer) peekPeekChar() byte {
	if l.readPosition+1 >= len(l.input) {
		return 0
	}
	return l.input[l.readPosition+1]
}

func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
		l.readChar()
	}
}

func (l *Lexer) skipComment() {
	for l.ch != '\n' && l.ch != 0 {
		l.readChar()
	}
}

func (l *Lexer) readNumber() string {
	position := l.position
	for isDigit(l.ch) {
		l.readChar()
	}
	return l.input[position:l.position]
}

func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

var keywords = map[string]TokenType{
	"render":   TOKEN_RENDER,
	"function": TOKEN_FUNCTION,
	"return":   TOKEN_RETURN,
	"print":    TOKEN_PRINT,
	"let":      TOKEN_LET,
	"import":   TOKEN_IMPORT,
	"before":   TOKEN_BEFORE,
	"string":   TOKEN_STRING,
	"int":      TOKEN_IDENT,
	"bool":     TOKEN_IDENT,
	"float":    TOKEN_IDENT,
	"void":     TOKEN_IDENT,
	"any":      TOKEN_IDENT,
	"match":    TOKEN_MATCH,
	"fn":       TOKEN_FN,
	"spawn":    TOKEN_SPAWN,
	"yield":    TOKEN_YIELD,
	"enum":     TOKEN_ENUM,
	"readonly": TOKEN_READONLY,
	"default":  TOKEN_IDENT,
	"if":       TOKEN_IF,
	"else":     TOKEN_ELSE,
	"while":    TOKEN_WHILE,
	"error":    TOKEN_IDENT,
	"SELECT":   TOKEN_IDENT,
	"FROM":     TOKEN_IDENT,
	"WHERE":    TOKEN_IDENT,
}

func lookupIdent(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return TOKEN_IDENT
}

File: src/core/lexer.rs
use logos::Logos;

#[derive(Logos, Debug, PartialEq, Clone)]
#[logos(skip r"[ \t\n\f]+")]
pub enum TokenType {
    #[token("render")]
    Render,
    #[token("function")]
    Function,
    #[token("return")]
    Return,
    #[token("print")]
    Print,
    #[token("println")]
    Println,
    #[token("let")]
    Let,
    #[token("import")]
    Import,
    #[token("before")]
    Before,
    #[token("string")]
    StringType,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("while")]
    While,
    #[token("for")]
    For,
    #[token("in")]
    In,
    #[token("match")]
    Match,
    #[token("route")]
    Route,
    #[token("GET")]
    Get,
    #[token("POST")]
    Post,
    #[token("PUT")]
    Put,
    #[token("DELETE")]
    Delete,
    #[token("PATCH")]
    Patch,
    #[token("query")]
    Query,
    #[token("spawn")]
    Spawn,
    #[token("yield")]
    Yield,
    #[token("enum")]
    Enum,
    #[token("readonly")]
    Readonly,
    #[token("test")]
    Test,
    #[token("struct")]
    Struct,
    #[token("try")]
    Try,
    #[token("catch")]
    Catch,
    #[token("finally")]
    Finally,

    #[token("int")]
    IntType,
    #[token("bool")]
    BoolType,
    #[token("float")]
    FloatType,
    #[token("void")]
    VoidType,
    #[token("any")]
    AnyType,
    #[token("default")]
    Default,
    #[token("<=")]
    Leq,
    #[token(">=")]
    Geq,
    #[token("error")]
    Error,
    #[token("SELECT")]
    Select,
    #[token("FROM")]
    From,
    #[token("WHERE")]
    Where,
    #[token("INSERT")]
    Insert,
    #[token("UPDATE")]
    Update,
    #[token("INTO")]
    Into,
    #[token("VALUES")]
    Values,
    #[token("SET")]
    Set,

    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*")]
    Ident,

    #[regex(r"\$[a-zA-Z_][a-zA-Z0-9_]*")]
    Var,

    #[regex(r"[0-9]+")]
    Int,

    #[regex(r"[0-9]+\.[0-9]+")]
    Float,

    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[token(".")]
    Dot,
    #[token("*")]
    Asterisk,
    #[token("/")]
    Slash,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    #[token("#!")]
    HashBang,
    #[token("#[")]
    LBracketHash,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("<")]
    Lt,
    #[token(">")]
    Gt,
    #[token("==")]
    Eq,
    #[token("!=")]
    NotEq,
    #[token("?")]
    Question,
    #[token("??")]
    Coalesce,
    #[token("!")]
    Bang,
    #[token("=")]
    Assign,
    #[token(":")]
    Colon,
    #[token("|>")]
    Pipe,
    #[token("!>")]
    Sanitize,
    #[token("?->")]
    Nullsafe,
    #[token("=>")]
    Arrow,
    #[token("&&")]
    And,
    #[token("||")]
    Or,
    #[token("@")]
    At,
    #[token("%")]
    Modulo,

    #[regex(r#""([^"\\]|\\.)*""#, |lex| lex.slice()[1..lex.slice().len()-1].to_string())]
    #[regex(r#"'([^'\\]|\\.)*'"#, |lex| lex.slice()[1..lex.slice().len()-1].to_string())]
    Literal(String),

    #[regex(r"//.*", logos::skip)]
    Comment,

    #[regex(r"/\*([^*]|\*[^/])*\*/", logos::skip)]
    MultiLineComment,

    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
    pub span: logos::Span,
}

pub struct Lexer<'a> {
    lexer: logos::Lexer<'a, TokenType>,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Lexer {
            lexer: TokenType::lexer(input),
        }
    }

    pub fn next_token(&mut self) -> Token {
        let token_type = self
            .lexer
            .next()
            .unwrap_or(Ok(TokenType::Eof))
            .unwrap_or(TokenType::Eof);
        let span = self.lexer.span();
        let literal = match &token_type {
            TokenType::Literal(s) => s.clone(),
            _ => self.lexer.slice().to_string(),
        };
        Token {
            token_type,
            literal,
            span,
        }
    }
}
